### PR TITLE
[Enhancement]  Dockerfile cleanup

### PR DESCRIPTION
**Pull request by amarburg**
## Changes Made

An opinionated cleanup of the dockerfile.  Includes:

- In the "ci" image:
    - Move the apt-get above the "copy" so changes to this repo don't break that cache line
    - Only copy */package.xml (requires the [recent "copy --parent" syntax](https://docs.docker.com/build/dockerfile/release-notes/#labs) in the Dockerfile) and blue.repos
    - `vcs import < blue.repos` so we rosdep dependencies for those packages into the image as well.

- In the "robot" image:
    - Move the gstreamer apt-get above the "copy" so changes to this repo don't break that cache

- Sort lists of packages for apt-get (https://docs.docker.com/build/building/best-practices/#sort-multi-line-arguments)

At present, this PR will not build due to other issues (#220 #223 ).  As this is an optimization, deal with this once the other issues have been resolved.

## Associated Issues

n/a

## Testing

None yet.

### CODE REVIEW

File: .docker/Dockerfile
Comment (evan-palmer) [inline .docker/Dockerfile:1]: This is neat. I hadn't seen this. When we merge, we should add an issue to track when that feature becomes stable, so that we can use the stable syntax

File: .docker/Dockerfile
Comment (evan-palmer) [inline .docker/Dockerfile:30]: The reason that I had this in the `robot` stage initially was to keep this image relatively lightweight for CI (the tests don't depend on the packages from `blue.repos`). I'm open to hearing your reasoning for the change though

File: .docker/Dockerfile
Comment (evan-palmer) [inline .docker/Dockerfile:37]: Same comment as above

**Review by evan-palmer — CHANGES_REQUESTED** [review]
Thanks for submitting this! I just have a couple of small comments. I saw that this was a draft after adding some comments, so I'll hold off on future reviews until it's ready for review.

File: .docker/Dockerfile
Comment (amarburg) [reply .docker/Dockerfile:1]: It was new to me as well.  I assume it requires a recent version of docker (or docker-compose plugin);  since builds happen in the Github action (or on our desktops), I don't think there are concerns with supporting end users with older Docker installs.

Agreed;  I note there are two newer Docker compose specs (1.8.0 and 1.9.0), but I don't know if this feature has been integrated into those, nor what versions of Docker are required to access those specs.

File: .docker/Dockerfile
Comment (amarburg) [reply .docker/Dockerfile:30]: This change specifically is a simple replacement for the previous "COPY blue ."  It only copies the package.xml files from "blue/*" rather than all of the not-docker-ignored files in Blue to reduce the dependencies which may break cache.

Address your question below...

File: .docker/Dockerfile
Comment (amarburg) [reply .docker/Dockerfile:37]: This is the critical section.  Essentially should CI contain (A) "rosdeps of blue" or (B) "rosdeps of blue + blue's deps"

There are two criteria:

1.  Does (A) exercise some test case in CI which (B) doesn't?   The only case I can think of is if we proactively install apt install a package which is _not correctly declared_ as a `package.xml` dependency in a `blue.repos` dependency.     I think that only happens if we remove a repo from `blue.repos` and don't regenerate the CI image.   That is:

- `blue.repos` includes package A, which requires `libfoo-dev`, so it's installed in the CI image
- Updated `blue.repos` to remove A but add B which requires `libfoo-dev` but erroneously does not require it in `package.xml`
- If we fail to rebuild the CI image, CI will succeed (incorrectly)
- If we rebuild the CI image, `libfoo-dev` will not be installed, and CI will fail (as expected)

2. How does it affect the size of the CI image?

As an example for 'iron':

- Without `blue.repos` CI image is  2.75GB
- With `blue.repos` CI image is  2.72GB

During testing with iron, I noticed this installed `mavros` (and dependencies) from package.   This may require some adjustment in #220

Comment (amarburg) [reply .docker/Dockerfile:37]: This is the critical section.  Essentially should CI contain (A) "rosdeps of blue" or (B) "rosdeps of blue + blue's deps"

Two potential criteria:

1.  Does (A) exercise some test case in CI which (B) doesn't?   The only case I can think of is if we proactively apt install a package which is _not correctly declared_ as a `package.xml` dependency in a `blue.repos` dependency.     I think that only happens if we remove a repo from `blue.repos` and then don't regenerate the CI image.   That is:

- `blue.repos` includes package X, which requires `libfoo-dev`, so it's installed in the CI image
- Updated `blue.repos` to remove X but add Y which requires `libfoo-dev` but erroneously does not require it in `package.xml`
- If we fail to rebuild the CI image, CI will succeed (incorrectly)
- If we rebuild the CI image, `libfoo-dev` will not be installed, and CI will fail (as expected)

2. How does it affect the size of the CI image?

As an example for 'iron':

- Without `blue.repos` CI image is  2.67 GB
- With `blue.repos` CI image is  2.72 GB

During testing with iron, I noticed this installed `mavros` (and dependencies) from package.   This may require some adjustment in #220 for rolling/jazzy.

Comment (evan-palmer) [reply .docker/Dockerfile:37]: Ah yes, you're completely right about point 1. I've been convinced :smile:  In fact, I think I remember removing the packages in the `blue.repos` file from the respective `package.xml` files because of this.

Comment (evan-palmer) [reply .docker/Dockerfile:37]: I'm going to add an issue once this is ready and merged mentioning adding the packages from `blue.repos` into the relevant `package.xml` files.

Comment (evan-palmer) [reply .docker/Dockerfile:37]: Issue documented in #249

**Comment by amarburg** [comment]
FYI this PR has become a bit gnarly due to upstream changes.   May be a hard push or a new PR in the future.

**Pull request closed without merging**
